### PR TITLE
Refactor common client TLS configuration.

### PR DESCRIPTION
**Pull request by ccaraman**
**Description:** 
NOTE: This does break existing behavior in a few ways:
1. It changes the config names for client TLS in gRPC scenarios.
ie `cert_pem_file` -> `ca_file`
`secure` -> ! `insecure`
2. Prior to this PR gRPC client connections were by default insecure connections. With this change, gRPC connections are by default secure. If an insecure connection is desired, the config option must be specified
` insecure:true` 

**Link to tracking Issue:** This is the first PR to address issue #933 
The next PR will update SecureReceiverSettings and address TLSServerConfig unification.

**Testing:**  This is mainly a refactoring PR so used existing tests.

**Documentation:**  Will update the documentation for impacted components after signoff of breaking change behavior.

### CODE REVIEW

**Comment by jrcamp** [comment]
Thanks so much for doing this! Looks like you'll need to rebase. For gRPC is it possible to use TLS but _not_ verify certificates? Or are the only options TLS on (with verify) or TLS off?

Minor nit: Just realized I think we could shorten `UseInsecure` to just `Insecure` as I don't think `Use` provides any additional information. But I can live with it either way. 😄

**Comment by ccaraman** [comment]
@jrcamp - see reply below 
> Thanks so much for doing this! Looks like you'll need to rebase.
Done 
> For gRPC is it possible to use TLS but _not_ verify certificates? Or are the only options TLS on (with verify) or TLS off?
Someone is going to have to look into that  - I'm not as familiar with gRPC TLS. As the refactor is, it just moved the existing behavior around.

> Minor nit: Just realized I think we could shorten `UseInsecure` to just `Insecure` as I don't think `Use` provides any additional information. But I can live with it either way. 😄
I have no preference. Can you add a comment in the config part itself to see if people have any strong opinions?

File: config/configtls/configtls.go [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+type TLSClientConfig struct {
+	TLSConfig `mapstructure:",squash"` // squash ensures fields are correctly decoded in embedded struct
+
+	// These are config options specific to client connections
+	// In gRPC when set to true, this is used to disable the client transport security.
+	// See https://godoc.org/google.golang.org/grpc#WithInsecure.
+	// In HTTP, this disables verifying the server's certificate chain and host name
+	// (InsecureSkipVerify in the tls Config).
+	// This is only used for client connections. (optional, default false)
+	UseInsecure bool `mapstructure:"insecure"`
```
Comment (jrcamp) [inline config/configtls/configtls.go:51]: nit: Would suggest just calling this `Insecure` like the config name, I don't think the `Use` adds any extra information.

File: config/configtls/configtls.go [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+	// Path to the TLS cert to use for TLS required connections. (optional)
+	CertFile string `mapstructure:"cert_file"`
+	// Path to the TLS key to use for TLS required connections. (optional)
+	KeyFile string `mapstructure:"key_file"`
+}
+
+// TLSClientConfig contains TLS configurations that are specific to client
+// connections in addition to the common configurations. This should be used by
+// components configuring TLS client connections.
+type TLSClientConfig struct {
```
Comment (bogdandrutu) [inline config/configtls/configtls.go:42]: I would use `TLSClientSettings`. It seems that we use as suffix `Config` for interfaces and `Settings` for concrete structs (or at least that is my impression, I may be wrong). But also the name is consistent with grpc :)

File: config/configtls/configtls.go [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+	"crypto/x509"
+	"fmt"
+	"io/ioutil"
+	"path/filepath"
+)
+
+// TLSConfig exposes the common client and server TLS configurations.
+// Note: Since there isn't anything specific to a server connection. Components
+// with server connections should use TLSConfig.
+type TLSConfig struct {
```
Comment (bogdandrutu) [inline config/configtls/configtls.go:28]: s/TLSConfig/TLSSettings? See motivation in the next comment.

File: config/configgrpc/configgrpc.go [hunk 4]
```diff
@@ -102,13 +80,13 @@ func GrpcSettingsToDialOptions(settings GRPCSettings) ([]grpc.DialOption, error)
...
-	if settings.TLSConfig.CaCert != "" && !settings.TLSConfig.UseSecure {
-		creds, err := credentials.NewClientTLSFromFile(settings.TLSConfig.CaCert, settings.TLSConfig.ServerNameOverride)
+	if settings.TLSConfig.CAFile != "" && settings.TLSConfig.UseInsecure {
+		creds, err := credentials.NewClientTLSFromFile(settings.TLSConfig.CAFile, settings.TLSConfig.ServerName)
 		if err != nil {
 			return nil, err
 		}
 		opts = append(opts, grpc.WithTransportCredentials(creds))
-	} else if settings.TLSConfig.UseSecure {
+	} else if !settings.TLSConfig.UseInsecure {
```
Comment (bogdandrutu) [inline config/configgrpc/configgrpc.go:89]: I would expect the logic around arguments in `TLSConfig` to be in a function in `TLSConfig` to ensure separation of concerns, I may be wrong but can we try to see if that is possible?

Comment (bogdandrutu) [reply config/configgrpc/configgrpc.go:89]: Also this code can be simplified (I personally see this more readable):
```golang
if settings.TLSConfig.UseInsecure {
  if settings.TLSConfig.CAFile != "" {

  } else {

  }
} else {

}
```

File: config/configtls/configtls.go (hunk 3 above)
Comment (ccaraman) [reply config/configtls/configtls.go:28]: Agreed that makes more sense. Will update.

File: config/configgrpc/configgrpc.go (hunk 4 above)
Comment (ccaraman) [reply config/configgrpc/configgrpc.go:89]: Two good points- I'll clean this up.

**Comment by ccaraman** [comment]
@bogdandrutu - it is ready for another review I addressed your comments! Thank you

**Pull request merged**